Mathlib/Combinatorics/Optimization/ValuedCSP.lean

```diff
@@ -65,3 +67,156 @@ def ValuedCsp.Instance.evalSolution {Γ : ValuedCsp D C} {ι : Type*}
 def ValuedCsp.Instance.IsOptimumSolution {Γ : ValuedCsp D C} {ι : Type*}
     (I : Γ.Instance ι) (x : ι → D) : Prop :=
   ∀ y : ι → D, I.evalSolution x ≤ I.evalSolution y
+
+def Function.HasMaxCutPropertyAt (f : (Fin 2 → D) → C) (a b : D) : Prop :=
+  f ![a, b] = f ![b, a] ∧  -- `argmin f` is exactly `{ ![a, b] , ![b, a] }`
+    ∀ x y : D, f ![a, b] ≤ f ![x, y] ∧ (f ![a, b] = f ![x, y] → a = x ∧ b = y ∨ a = y ∧ b = x)
+
+def Function.HasMaxCutProperty (f : (Fin 2 → D) → C) : Prop :=
+  ∃ a b : D, a ≠ b ∧ f.HasMaxCutPropertyAt a b
+
+def FractionalOperation (D : Type _) (m k : ℕ) : Type* :=
+  (Fin m → D) → (Fin k → D)
+
+/-
+For documentation purposes, here is a "less compact" version of the upcoming two definitions:
+
+[import Mathlib.Data.Matrix.Basic]
+
+def FractionalOperation.tt {D : Type} {m k n : ℕ}
+    (ω : FractionalOperation D m k) (x : (Fin m → (Fin n → D))) :
+    (Fin k → (Fin n → D)) :=
+  Matrix.transpose (fun (i : Fin n) => ω ((Matrix.transpose x) i))
+
+def Function.AdmitsFractional {D C : Type} [OrderedAddCommMonoid C] {n m k : ℕ}
+    (f : (Fin n → D) → C) (ω : FractionalOperation D m k) : Prop :=
+  ∀ x : (Fin m → (Fin n → D)),
+    m • List.sum (List.map (fun i => f ((ω.tt x) i)) (List.finRange k)) ≤
+    k • List.sum (List.map (fun i => f (      x  i)) (List.finRange m))
+
+They should be defeq to the definitions below.
+-/
+
+def FractionalOperation.tt {m k n : ℕ}
+    (ω : FractionalOperation D m k) (x : (Fin m → (Fin n → D)))
+    (a : Fin k) (b : Fin n) : D :=
+  (ω (fun (i : Fin m) => x i b)) a
```

Comment:
Are these intended to be matrices?

Reply:
I mean, they are several "vectors". I don't really perform typical linear-algebraic matrix manipulations with them. Nevertheless, the operation `.tt` can be expressed in the language of matrix transpositions plus extra steps (see the comment above this definition).

Reply:
Use the matrix notation: `!![a,b;c,d]` (maybe with comma and semicolon swapped).

Reply:
Should I also use `Matrix` as type?

Reply:
You should only use `!!` notation if you are using `Matrix` as a type. Whether you use `Matrix` will depend on whether mathematically you want to think of these as matrices, or tuples of tuples.

Reply:
If you genuinely would never multiply them (or dot product with a vector), then I agree that `Matrix` is probably a bad idea.